### ISSUE TITLE
Random floating-point errors in GitHub Actions

### ISSUE THREAD

**Issue by dsgibbons** [shap/shap#2960]
Copied from https://github.com/dsgibbons/shap/issues/98

We have seen GitHub Actions randomly fail occasionally due to floating-point precision issues. See, for example:

- https://github.com/dsgibbons/shap/actions/runs/5085485586/attempts/1 (EDIT: py3.8, `tests/explainers/test_deep.py::test_pytorch_mnist_cnn`)
- https://github.com/dsgibbons/shap/actions/runs/5135384415/attempts/1 (EDIT: py3.11, `tests/explainers/test_deep.py::test_pytorch_single_output`)

In both cases, the tests passed after being re-run with no source code changes. We should consider either increasing some of the floating-point tolerances, or investigate why we are seeing such large random deviations in `test_deep.py`.

## List of flaky tests

Here's a list of all the flaky tests we know of so far (feel free to edit). 

- [ ] `test_tf_keras_linear` with seed `256`
- [ ] `test_single_tree_compare_with_kernel_shap` with seed `434`
- [ ] `test_single_tree_nonlinear_transformations` with seed `116`
- [ ] `test_xgboost_classifier_independent_margin` with seed `364`
- [ ] `test_gradient.py::test_pytorch_mnist_cnn` with seed `24`
- [ ] `test_deep.py::test_pytorch_mnist_cnn` with seed `212`
- [ ] `test_pytorch_single_output`
- [ ] `test_xgboost_direct` with seed `28`
- [ ] `test_several_trees` with seed `411`
- [ ] `test_shape_values_linear_many_features` with seed `94` (**this test seems particular egregious, it's failing 59 out of the first 100 seeds**)

**Comment by dsgibbons** [comment]
Another example, this time from `test_tree.py`: https://github.com/slundberg/shap/actions/runs/5163671345/attempts/1

(EDIT: py3.8, `tests/explainers/test_tree.py::test_xgboost_direct`)

**Comment by thatlittleboy** [comment]
I came across a tip that is helpful in debugging this, for those interested in tackling this. https://calmcode.io/pytest-tricks/flaky-tests.html

For the `tests/explainers/test_tree.py::test_xgboost_direct` test, I managed to reproduce the failing test for seed=28.

```python
def test_xgboost_direct(i=28):
    np.random.seed(i)  # <----------------------------------
    xgboost = pytest.importorskip('xgboost')

    N = 100
    M = 4
    X = np.random.randn(N, M)
    y = np.random.randn(N)

    model = xgboost.XGBRegressor()
    model.fit(X, y)

    explainer = shap.TreeExplainer(model)
    shap_values = explainer.shap_values(X)

    assert np.allclose(shap_values[0, :], _brute_force_tree_shap(explainer.model, X[0, :]))
```

EDIT 1: an obvious "fix" would be to fix the random seed to 0 or something, but it would be nice to figure out why it fails sometimes.

**Comment by dsgibbons** [comment]
A seed is a good idea. Let's start implementing this whenever we notice random failures.

**Comment by thatlittleboy** [comment]
Apart from the xgboost failures, the `test_deep.py` tests (e.g. `test_pytorch_mnist_cnn`) are also failing quite frequently.

https://github.com/slundberg/shap/actions/runs/5373998736/jobs/9748916641

```
>       assert d / np.abs(diff).sum() < 0.001, "Sum of SHAP values does not match difference! %f" % (d / np.abs(diff).sum())
E       AssertionError: Sum of SHAP values does not match difference! 0.003544
E       assert (1.0529657197111408e-05 / 0.0029708666) < 0.001
E        +  where 0.0029708666 = <built-in method sum of numpy.ndarray object at 0x7fa011a522b0>()
E        +    where <built-in method sum of numpy.ndarray object at 0x7fa011a522b0> = array([6.20380044e-04, 1.16837524e-04, 2.74385005e-04, 3.91309470e-04,\n       3.74433905e-04, 6.30353868e-04, 8.25027601e-05, 2.51042336e-04,\n       9.09467562e-05, 1.38675168e-04], dtype=float32).sum
E        +      where array([6.20380044e-04, 1.16837524e-04, 2.74385005e-04, 3.91309470e-04,\n       3.74433905e-04, 6.30353868e-04, 8.25027601e-05, 2.51042336e-04,\n       9.09467562e-05, 1.38675168e-04], dtype=float32) = <ufunc 'absolute'>(array([-6.20380044e-04, -1.16837524e-04, -2.74385005e-04, -3.91309470e-04,\n        3.74433905e-04,  6.30353868e-04, -8.25027601e-05,  2.51042336e-04,\n        9.09467562e-05,  1.38675168e-04], dtype=float32))
E        +        where <ufunc 'absolute'> = np.abs

tests/explainers/test_deep.py:335: AssertionError
```

We have already set the np random seed for this test, so probably it's some other random seed we need to keep fixed. E.g. torch seed?

**Comment by connortann** [comment]
Here's a list of all the flaky tests we know of so far. EDIT: moved checklist up to main issue description. 

Once we have identified a test as being flaky, I propose that in the short term we pin the seed for _that_ test to something that passes whilst we track the issue here and diagnose the root cause (or just increase the tolerance).

I propose we tick them off as "fixed" once we know the test passes reliably for any random seed, e.g. with the `random_seed` fixture proposed in #3048. One way to run a test locally with a range of different seeds is to parameterise the test like:

```python
@pytest.mark.parametrize("random_seed", range(100))
```

Nb. in #3048 I've added a CLI argument to pytest that controls the value of the `random_seed` fixture, e.g.

```bash
pytest -k my_test --random-seed 123
```

**Comment by connortann** [comment]
I think perhaps that in some cases, fixing the seed might be the best long-term way forward - it probably depends on the nature of the test.

- I think the _default_ should probably be to have **completely deterministic** tests: that's a usual "best practice" for testing, and it's most developer's default assumption when running a suite.
- **Fuzz-testing** (i.e. changing inputs on each run) can be helpful to get better coverage. But, it should only be used for tests that we demand must pass for _any_ possible value of the randomised inputs.

Tests that evaluate the accuracy of a sampling-based estimation (e.g. checking an accuracy to within a floating-point tolerance) may _not_ have any guaranteed upper bound on the error, so there's always a nonzero probability the accuracy will be below any given bound. In that scenario, a fuzzed test will fail with nonzero probability, so the most useful test is probably one with a fixed seed.

An overarching principle is that any test should pass *reliably*, as flaky tests are annoying, hard to debug, and impede productivity. In terms of priorities, IMO "having a reliable CI suite" >> "better coverage on some tests". So, if we decide to add fuzz testing to a test which was previously failing, we should probably run it many times to ensure it's reliable (say, 100?)

**Comment by thatlittleboy** [comment]
Yep, I don't disagree with any of your points. I won't insist on fuzz testing for coverage, that's not my intended purpose.

As long as we are sure that the failing tests are NOT because of an algorithm/implementation error but just something we expect will arise from sampling, then I'm okay putting this issue to rest.

In which case, we can be confident replying to all of the prior bug issues (about addivity checks failing) that it's just a sampling issue and they can safely ignore the checks. And maybe we could also phrase the error / warning message a little more clearly to convey this point slightly better.

**Comment by connortann** [comment]
One way to investigate could be to record the magnitude of the discrepancy for a range of seeds, and look at the histogram. It would be interesting to see if its like say a Gamma distribution, or if it's multimodal or very fat tailed
